packages/zero-client/src/client/custom.ts

```diff
@@ -116,17 +118,35 @@ export class TransactionImpl<S extends Schema> implements ClientTransaction<S> {
   readonly token: string | undefined;
 }
 
-// TODO: wire into `PushResponseTracker` so we can give out promises for each mutation
-// when it is `initial`
+type MaybeWithServerResult<T> = {
+  server?: Promise<T>;
+};
+
+type PromiseMaybeWithServerResult<T, S> = Promise<T> & MaybeWithServerResult<S>;
+export type PromiseWithServerResult<T, S> = Promise<T> & {server: Promise<S>};
+
 export function makeReplicacheMutator<S extends Schema>(
   lc: LogContext,
+  mutationTracker: MutationTracker,
   mutator: CustomMutatorImpl<S>,
   schema: S,
   slowMaterializeThreshold: number,
 ) {
-  return (repTx: WriteTransaction, args: ReadonlyJSONValue): Promise<void> => {
+  return (
+    repTx: WriteTransaction,
+    args: ReadonlyJSONValue,
+  ): PromiseMaybeWithServerResult<void, MutationResult> => {
     const tx = new TransactionImpl(lc, repTx, schema, slowMaterializeThreshold);
-    return mutator(tx, args);
+    const clientPromise = mutator(tx, args);
+
+    if (repTx.reason === 'initial') {
```

Comment:
note to self: add some test to verify we do not try to track mutations on a rebase.